# Assignment 2: Linear Models and Validation Metrics (30 marks total)
### Due: October 10 at 11:59pm

### Name: 

### In this assignment, you will need to write code that uses linear models to perform classification and regression tasks. You will also be asked to describe the process by which you came up with the code. More details can be found below. Please cite any websites or AI tools that you used to help you with this assignment.

## Part 1: Classification (14.5 marks total)

You have been asked to develop code that can help the user determine if the email they have received is spam or not. Following the machine learning workflow described in class, write the relevant code in each of the steps below:

### Step 0: Import Libraries

In [217]:
import numpy as np
import pandas as pd
from yellowbrick.datasets.loaders import load_spam
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score


### Step 1: Data Input (1 mark)

The data used for this task can be downloaded using the yellowbrick library: 
https://www.scikit-yb.org/en/latest/api/datasets/spam.html

Use the yellowbrick function `load_spam()` to load the spam dataset into the feature matrix `X` and target vector `y`.

Print the size and type of `X` and `y`.

In [218]:
# TO DO: Import spam dataset from yellowbrick library
X, y = load_spam()
# TO DO: Print size and type of X and y
print("Shape of X:", X.shape)
print("Type of X:", type(X))
print("Shape of y:", y.shape)
print("Type of y:", type(y))

Shape of X: (4600, 57)
Type of X: <class 'pandas.core.frame.DataFrame'>
Shape of y: (4600,)
Type of y: <class 'pandas.core.series.Series'>


### Step 2: Data Processing (1.5 marks)

Check to see if there are any missing values in the dataset. If necessary, select an appropriate method to fill-in the missing values.

In [219]:
# TO DO: Check if there are any missing values and fill them in if necessary
# I need to get the summary of the dataset. This gives an overview of Non-Null count
# There are 57 columns and 4600 rowa
X.info()
y.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4600 entries, 0 to 4599
Data columns (total 57 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   word_freq_make              4600 non-null   float64
 1   word_freq_address           4600 non-null   float64
 2   word_freq_all               4600 non-null   float64
 3   word_freq_3d                4600 non-null   float64
 4   word_freq_our               4600 non-null   float64
 5   word_freq_over              4600 non-null   float64
 6   word_freq_remove            4600 non-null   float64
 7   word_freq_internet          4600 non-null   float64
 8   word_freq_order             4600 non-null   float64
 9   word_freq_mail              4600 non-null   float64
 10  word_freq_receive           4600 non-null   float64
 11  word_freq_will              4600 non-null   float64
 12  word_freq_people            4600 non-null   float64
 13  word_freq_report            4600 

For this task, we want to test if the linear model would still work if we used less data. Use the `train_test_split` function from sklearn to create a new feature matrix named `X_small` and a new target vector named `y_small` that contain **5%** of the data.

In [220]:
#create a new feature matrix named X_small and target vector names y_small that contains 5% of the data

X_small_train, X_small_test, y_small_train, y_small_test, = train_test_split(X, y, test_size = 0.95, random_state = 0)

### Step 3: Implement Machine Learning Model

1. Import `LogisticRegression` from sklearn
2. Instantiate model `LogisticRegression(max_iter=2000)`.
3. Implement the machine learning model with three different datasets: 
    - `X` and `y`
    - Only first two columns of `X` and `y`
    - `X_small` and `y_small`

### Step 4: Validate Model

Calculate the training and validation accuracy for the three different tests implemented in Step 3

### Step 5: Visualize Results (4 marks)

1. Create a pandas DataFrame `results` with columns: Data size, training accuracy, validation accuracy
2. Add the data size, training and validation accuracy for each dataset to the `results` DataFrame
3. Print `results`

In [224]:
X_subset = X.iloc[:, :2]
X_subset.describe()
y_subset = y
y_subset

0       1
1       1
2       1
3       1
4       1
       ..
4595    0
4596    0
4597    0
4598    0
4599    0
Name: is_spam, Length: 4600, dtype: int64

In [226]:
# TO DO: ADD YOUR CODE HERE FOR STEPS 3-5
# Note: for any random state parameters, you can use random_state = 0
# HINT: USING A LOOP TO STORE THE DATA IN YOUR RESULTS DATAFRAME WILL BE MORE EFFICIENT
# Instantiate model LogisticRehression with maximum iteration of 2000
#logisticRegression = LogisticRegression(max_iter=2000)

# Create an empty DataFrame
results = pd.DataFrame(columns=["Data size", "Training accuracy", "Validation accuracy"])

# List to store the sizes of the three training datasets
data_sizes = []

# List to store the training accuracies for the three datasets
training_accuracies = []

# List to store the validation accuracies for the three datasets
validation_accuracies = []


# implement the machine learning model with three different datasets

# make X_subset the first two columns of X and y

#X_subset = X.iloc[:, :2]
#y_subset = y

#Using the first dataset X and y instantiate a logisticRegression object
logisticRegression_original = LogisticRegression(max_iter=2000) # create an instance of the logisticRegression model

#split into training and test sets by given test size of 5%
X_train_original, X_test_original, y_train_original, y_test_original = train_test_split(X, y, test_size = 0.05, random_state = 0)

data_sizes.append(len(X_train_original))

#train the model
logisticRegression_original.fit(X_train_original, y_train_original)

# predict the value of y on the X_test set
y_prediction_original_test = logisticRegression_original.predict(X_test_original)
y_prediction_original_train = logisticRegression_original.predict(X_train_orginal)

# Only the first two columns of X and y
X_subset_train, X_subset_test, y_subset_train, y_subset_test = train_test_split(X_subset, y_subset, test_size = 0.05, random_state = 0)
logisticRegression_subset = LogisticRegression(max_iter=2000)
logisticRegression_subset.fit(X_subset_train, y_subset_train)

data_sizes.append(len(X_subset_train))

# prediction on the X_subset test size
y_subset_column_prediction_test = logisticRegression_subset.predict(X_subset_test)
y_subset_column_prediction_train = logisticRegression_subset.predict(X_subset_train)

#implementation of the model on the X_small and y_small training set
logisticRegression_small = LogisticRegression(max_iter = 2000)
logisticRegression_small.fit(X_small_train, y_small_train)

data_sizes.append(len(X_small_train))
# prediction on the X_small training and X_small test set
y_small_prediction_test = logisticRegression_small.predict(X_small_test)
y_small_prediction_train = logisticRegression_small.predict(X_small_train)

# To calculate the training and validation accuracy for the three different sets

#Let us consider the original dataset X, y
training_accuracies.append(accuracy_score(y_train_original, y_prediction_original_train))
validation_accuracies.append(accuracy_score(y_test_original, y_prediction_original_test))


# Let us consider the subset of 2 columns

training_accuracies.append(accuracy_score(y_subset_train, y_subset_column_prediction_train))
validation_accuracies.append(accuracy_score(y_subset_test, y_subset_column_prediction_test))


#let us consider the X_small and y_small

training_accuracies.append(accuracy_score(y_small_train, y_small_prediction_train))
validation_accuracies.append(accuracy_score(y_small_test, y_small_prediction_test))

# Add the results to the DataFrame
results["Data size"] = data_sizes
results["Training accuracy"] = training_accuracies
results["Validation accuracy"] = validation_accuracies

# Print the results DataFrame
print(results)




   Data size  Training accuracy  Validation accuracy
0       4370           0.518764             0.943478
1       4370           0.616247             0.578261
2        230           0.947826             0.905034


### Questions (4 marks)
1. How do the training and validation accuracy change depending on the amount of data used? Explain with values.
2. In this case, what do a false positive and a false negative represent? Which one is worse?

*YOUR ANSWERS HERE*
In this case, for dataset with small sample , there was a similarity between the validation accuracy and the training accuracy. The deviation from both values was about 0.03 but as the samples increased, the training acccuracy seem to dip significantly.
It seems there must have been a lot of underfitting. There seems not to be a complex similarities between the features of the data. When unseen data was used during the validation, the accuracy seems to improve significantly.when I used more data at 95% for the training and 5% for testing but faltered when I used only 5% of the data for training and used the outcome on the test data at 95%.

False positives occurs when the model predicts a positive output instead of the true output which is negative
False negatives occurs when the model predicts a negative output when the true output should have been positive.
The situation and reason for the analysis will determine with of the scenrio is preferred.

### Process Description (4 marks)
Please describe the process you used to create your code. Cite any websites or generative AI tools used. You can use the following questions as guidance:
1. Where did you source your code?
1. In what order did you complete the steps?
1. If you used generative AI, what prompts did you use? Did you need to modify the code at all? Why or why not?
1. Did you have any challenges? If yes, what were they? If not, what helped you to be successful?

*DESCRIBE YOUR PROCESS HERE*
I followed the algorithm as described in the lecture note provided in the classroom on D2L
What I had to do was understand the data.I imported the data using the guide I got by learning the tutorial [here](https://scikit-learn.org/stable/tutorial/basic/tutorial.html)
Review the data : I understood the type of data and the column heads. I checked if there was any null function in any of the rows or columns.
Since I could see the data was clean, I proceeded to divide the dataset into 3 according to the instruction. The orginal set with 5% test data and 95% train data, the second data was 5% train data and 95% test data while the 
third data was 2 column subset of the original data.
I split the data using the train_test_split function from the sklearn model selection module library to get the split asa i  described it. The random size was left at 0 for consistency of the randomization.
I imported the linear regression model from the sklearn module library and passed the training set I had split to it. I used the fit function to train the model and used the predict function to verify the learning model worked.
Confirmation was done by calculating the validation and train accuracy of the data. This is compared to further look into the outcome for analysis.

Challenges I faced includes understanding and revisiting some Numpy and panda codes for data retrieval and looping. I also had to use the scikit learn website a lot to understand and retrieve the codes to execute the different stages of the process.
I needed to learn underfitting and overfitting to undertand the comparison and differences in the validation and training accuracy. Constant learning and practice is required to get good at machine learning. A deep understanding of the model itself will be necessary for success.

## Part 2: Regression (10.5 marks total)

For this section, we will be evaluating concrete compressive strength of different concrete samples, based on age and ingredients. You will need to repeat the steps 1-4 from Part 1 for this analysis.

### Step 1: Data Input (1 mark)

The data used for this task can be downloaded using the yellowbrick library: 
https://www.scikit-yb.org/en/latest/api/datasets/concrete.html

Use the yellowbrick function `load_concrete()` to load the spam dataset into the feature matrix `X` and target vector `y`.

Print the size and type of `X` and `y`.

In [281]:
# Import spam dataset from yellowbrick library
from yellowbrick.datasets.loaders import load_concrete
from sklearn.metrics import mean_squared_error, r2_score
from sklearn.linear_model import LinearRegression

X, y = load_concrete()

# TO DO: Print size and type of X and y
print("Shape of X:", X.shape)
print("Type of X:", type(X))
print("Shape of y:", y.shape)
print("Type of y:", type(y))

Shape of X: (1030, 8)
Type of X: <class 'pandas.core.frame.DataFrame'>
Shape of y: (1030,)
Type of y: <class 'pandas.core.series.Series'>


### Step 2: Data Processing (0.5 marks)

Check to see if there are any missing values in the dataset. If necessary, select an appropriate method to fill-in the missing values.

In [282]:
# TO DO: Check if there are any missing values and fill them in if necessary
#Get a summary of all the data in X
K = X.describe()
print(K)
#check for missing values in feature matrix X
missing_null_X = X.isnull()
#print(missing_null_X)
#check for missing values or null in the target vector y
missing_null_y = y.isnull()
#print(missing_null_y)
#count the missing values in each column
missing_count = missing_values.sum()
#print(missing_count)

            cement         slag          ash        water       splast  \
count  1030.000000  1030.000000  1030.000000  1030.000000  1030.000000   
mean    281.165631    73.895485    54.187136   181.566359     6.203112   
std     104.507142    86.279104    63.996469    21.355567     5.973492   
min     102.000000     0.000000     0.000000   121.750000     0.000000   
25%     192.375000     0.000000     0.000000   164.900000     0.000000   
50%     272.900000    22.000000     0.000000   185.000000     6.350000   
75%     350.000000   142.950000   118.270000   192.000000    10.160000   
max     540.000000   359.400000   200.100000   247.000000    32.200000   

            coarse         fine          age  
count  1030.000000  1030.000000  1030.000000  
mean    972.918592   773.578883    45.662136  
std      77.753818    80.175427    63.169912  
min     801.000000   594.000000     1.000000  
25%     932.000000   730.950000     7.000000  
50%     968.000000   779.510000    28.000000  
75% 

### Step 3: Implement Machine Learning Model (1 mark)

1. Import `LinearRegression` from sklearn
2. Instantiate model `LogisticRegression(max_iter=2000)`.
3. Implement the machine learning model with `X` and `y`

In [285]:
# TO DO: ADD YOUR CODE HERE
# Note: for any random state parameters, you can use random_state = 0

# Under consideration will be the full dataset, a small dataset which will be 5% data of the original dataset and a two column only of the original dataset
# I will create a loop to execute this task for each of the dataset

# create a dataset that only contains two columns of the original dataset

X_TwoColumn = X.iloc[:, :2]

# using train_test_split function, create the different subset of the original dataset
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.05, random_state = 0)

# using train_test_split function to use just 5% of the original dataset
X_5_percentData_train, X_5_percentData_test, y_5_percentData_train, y_5_percentData_test = train_test_split(X, y, test_size = 0.95, random_state = 0)

X_TwoColumn_train, X_TwoColumn_test, y_TwoColumn_train, y_TwoColumn_test = train_test_split(X_TwoColumn, y, test_size = 0.05, random_state = 0)

# Function to fit a model
def fit_model(X_train, y_train):
    model = LinearRegression()
    model.fit(X_train, y_train)
    return model

# Fit the model on the full dataset
model_X = fit_model(X_train, y_train)

# Fit the model on the 5% dataset
model_5_percentData = fit_model(X_5_percentData_train, y_5_percentData_train)

# Fit the model on the two-column dataset
model_TwoColumn = fit_model(X_TwoColumn_train, y_TwoColumn_train)



### Step 4: Validate Model (1 mark)

Calculate the training and validation accuracy using mean squared error and R2 score.

In [286]:
# Mean Squared Error (MSE) measures the average squared difference between the actual values and the predicted values. 
# Lower MSE values indicate better model performance.

# The R2 score quantifies the fraction of the variability in the target variable (dependent variable) that can be 
# anticipated based on the features (independent variables). It's a value between 0 and 1, 
# with 1 indicating that the model perfectly predicts the target variable and 0 indicating that the model doesn't explain any of the variance.

# Create an empty DataFrame
result = pd.DataFrame(columns=["MSE", "R2 Score"])

# Function to calculate MSE and R2 score
def calculate_metrics(model, X_test, y_test):
    y_test_pred = model.predict(X_test)
    MSE = mean_squared_error(y_test, y_test_pred)
    R2 = r2_score(y_test, y_test_pred)
    return MSE, R2

# calculate the MSE and R2 score for the full data test set
MSE_X, R2_X = calculate_metrics(model_X, X_test, y_test)
# calculate the MSE and R2 score for the full data train set
MSE_X_train, R2_X_train = calculate_metrics(model_X, X_train, y_train)

# Assign rows with the label to put data in the columns
result.loc["Full_Data_test"] = [MSE_X, R2_X]
result.loc["Full Data_train"] = [MSE_X_train, R2_X_train]

# Calculate the MSE and R2 score for the 5% dataset on test data
MSE_5_percentData_test, R2_5_percentData_test = calculate_metrics(model_5_percentData, X_5_percentData_test, y_5_percentData_test)
# Calculate the MSE and R2 score for the 5% dataset on training data
MSE_5_percentData_train, R2_5_percentData_train = calculate_metrics(model_5_percentData, X_5_percentData_train, y_5_percentData_train)

result.loc["5% Data_test"] = [MSE_5_percentData_test, R2_5_percentData_test]
result.loc["5% Data_train"] = [MSE_5_percentData_train, R2_5_percentData_train]


# Calculate the MSE and R2 score for the Two column test dataset 
MSE_TwoColumn_test, R2_TwoColumn_test = calculate_metrics(model_TwoColumn, X_TwoColumn_test, y_TwoColumn_test)
# Calculate the MSE and R2 score for the Two column dataset train data
MSE_TwoColumn_train, R2_TwoColumn_train = calculate_metrics(model_TwoColumn, X_TwoColumn_train, y_TwoColumn_train)

result.loc["Two-Column Data_test"] = [MSE_TwoColumn_test, R2_TwoColumn_test]
result.loc["Two-Column Data_train"] = [MSE_TwoColumn_train, R2_TwoColumn_train]



### Step 5: Visualize Results (1 mark)
1. Create a pandas DataFrame `results` with columns: Training accuracy and Validation accuracy, and index: MSE and R2 score
2. Add the accuracy results to the `results` DataFrame
3. Print `results`

In [287]:
# TO DO: ADD YOUR CODE HERE
# # Add the accuracy results to the results DataFrame
# transpose the frame to match the request

result = result.transpose()

print(result)

          Full_Data_test  Full Data_train  5% Data_test  5% Data_train  \
MSE           106.595189       107.313261    190.718486      90.347589   
R2 Score        0.636482         0.613545      0.304147       0.755427   

          Two-Column Data_test  Two-Column Data_train  
MSE                 196.639050             186.938708  
R2 Score              0.329408               0.326799  


### Questions (2 marks)
1. Did using a linear model produce good results for this dataset? Why or why not?

No. The linear regression model did not produce a better result than the logistic regression model. At all the datasets, the prediction seems to get worse on the actual data. There seems to be no clear pattern identification for the dataset.

### Process Description (4 marks)
Please describe the process you used to create your code. Cite any websites or generative AI tools used. You can use the following questions as guidance:
1. Where did you source your code?
1. In what order did you complete the steps?
1. If you used generative AI, what prompts did you use? Did you need to modify the code at all? Why or why not?
1. Did you have any challenges? If yes, what were they? If not, what helped you to be successful?

*DESCRIBE YOUR PROCESS HERE*
I source my code at different state of the execution by using the [scikit learn website](https://www.scikit-yb.org/en/latest/tutorial.html#feature-extraction) and lecture notes examples. I also used [youtube video](https://www.youtube.com/watch?v=29ZQ3TDGgRQ).The same process as used in my earlier process was used here again. I imported the model and viewed the data to review and ensure there is no missing values. I used describe() to count each data entry is consistent for each column.I then proceeded to split and train the model. I had to read about the defination of mean squared error and R2 score. This helped me interprete the result of the analysis.

## Part 3: Observations/Interpretation (3 marks)

Describe any pattern you see in the results. Relate your findings to what we discussed during lectures. Include data to justify your findings.


*ADD YOUR FINDINGS HERE*

For the full data set, I used 20% to test the model. The result between the mean squared error and the R2 score for both the training and test did not seem to vary at all. But when compared to when less data was used, the model performed worse wehn trying to predict a pattern as seen in the R2 score between the training accuracy and the validation accuracy columns.

## Part 4: Reflection (2 marks)
Include a sentence or two about:
- what you liked or disliked,
- found interesting, confusing, challangeing, motivating
while working on this assignment.


*ADD YOUR THOUGHTS HERE*
I liked that the process was defined step by step but it can be challenging trying to deal with missing data. Taking a decision about how to deal with missing data leaves a question about how that could have affected the outcome of the model. It can be a bit confusing trying to deal with how much of the data should be used for test and training the data. How can I know that the dataset used is not biased and free from noise. 

It is challenging going through the tutorial on scikit learn website if you are not mathematically inclined. The choice of what model to use has a foundation in mathematics and that is where the knowledge of mathematics will play a significant roll from the start when choosing what model to use for a specific type of datset.

It is very motivating to see how the model predicts the pattern and gets the outcome right.

## Part 5: Bonus Question (4 marks)

Repeat Part 2 with Ridge and Lasso regression to see if you can improve the accuracy results. Which method and what value of alpha gave you the best R^2 score? Is this score "good enough"? Explain why or why not.

**Remember**: Only test values of alpha from 0.001 to 100 along the logorithmic scale.

In [293]:
# TO DO: ADD YOUR CODE HERE
from sklearn.linear_model import Ridge, Lasso
from yellowbrick.datasets.loaders import load_concrete
from sklearn.metrics import mean_squared_error, r2_score

X, y = load_concrete()

X_train, X_test, y_train, y_test = train_test_split(X, y , test_size = 0.05, random_state = 0)

# Create a list of alpha values on a logarithmic scale
alphas = np.logspace(-3, 2, num=6)  # Values range from 0.001 to 100

for alpha in alphas:
    # Create Ridge and Lasso models with the current alpha value
    ridge_model = Ridge(alpha=alpha)
    lasso_model = Lasso(alpha=alpha)

    # Fit the models to the training data
    ridge_model.fit(X_train, y_train)  
    lasso_model.fit(X_train, y_train) 
    
    # Make predictions on validation data
    ridge_predictions_train = ridge_model.predict(X_train)
    lasso_predictions_train = lasso_model.predict(X_train)
    ridge_predictions_test = ridge_model.predict(X_test)
    lasso_predictions_test = lasso_model.predict(X_test)

# Evaluate performance (e.g., calculate MSE and R2 score)
    ridge_mse = mean_squared_error(y_train, ridge_predictions_train)
    ridge_r2 = r2_score(y_test, ridge_predictions_test)

    lasso_mse = mean_squared_error(y_train, lasso_predictions_train)
    lasso_r2 = r2_score(y_test, lasso_predictions_test)

    print(f"Ridge - Alpha: {alpha}, MSE: {ridge_mse}, R2 Score: {ridge_r2}")
    print(f"Lasso - Alpha: {alpha}, MSE: {lasso_mse}, R2 Score: {lasso_r2}")


Ridge - Alpha: 0.001, MSE: 107.31326063548332, R2 Score: 0.6364816968566105
Lasso - Alpha: 0.001, MSE: 107.31326069823315, R2 Score: 0.6364862921556714
Ridge - Alpha: 0.01, MSE: 107.31326063548407, R2 Score: 0.6364817081342449
Lasso - Alpha: 0.01, MSE: 107.3132669101057, R2 Score: 0.636527629748241
Ridge - Alpha: 0.1, MSE: 107.31326063555903, R2 Score: 0.6364818209091729
Lasso - Alpha: 0.1, MSE: 107.31388803830968, R2 Score: 0.6369382231639868
Ridge - Alpha: 1.0, MSE: 107.31326064305226, R2 Score: 0.6364829485172998
Lasso - Alpha: 1.0, MSE: 107.375946393138, R2 Score: 0.6407662627314561
Ridge - Alpha: 10.0, MSE: 107.31326139115107, R2 Score: 0.6364942104990654
Lasso - Alpha: 10.0, MSE: 109.13012460425415, R2 Score: 0.6544718019495821
Ridge - Alpha: 100.0, MSE: 107.31333499468136, R2 Score: 0.6366054362624083
Lasso - Alpha: 100.0, MSE: 148.77532864744654, R2 Score: 0.5172771180992553


*ANSWER HERE*
Lasso with an alpha of 10 gave me the best R2 score at 0.65 on a full data set with 80% training data and 20% test data.